# Common table expressions

Using the northwind database

In [1]:
import pymysql # pip install pymysql
import sqlalchemy as alch # bash: conda activate ironhack; python -m pip install --upgrade 'sqlalchemy<2.0'
                                                        # pip install SQLAlchemy
from getpass import getpass
import pandas as pd
import os
from getpass import getpass

In [2]:
password = getpass()

········


In [3]:
from getpass import getpass
password = getpass()
dbName = "northwind"
connectionData=f"mysql+pymysql://root:{password}@localhost/{dbName}"
engine = alch.create_engine(connectionData)

········


## 1. Write a CTE that lists the names and quantities of products with a unit price greater than $50.

In [18]:
query ="WITH ExpensiveProducts AS (SELECT ProductName, Unit FROM products WHERE Price > 50) SELECT ProductName, Unit FROM ExpensiveProducts;"
pd.read_sql_query(query, engine)


ProductName                  Unit
0          Mishi Kobe Niku      18 - 500 g pkgs.
1         Carnarvon Tigers            16 kg pkg.
2   Sir Rodney's Marmalade         30 gift boxes
3  Thüringer Rostbratwurst  50 bags x 30 sausgs.
4            Côte de Blaye    12 - 75 cl bottles
5    Manjimup Dried Apples      50 - 300 g pkgs.
6     Raclette Courdavault             5 kg pkg.

```sql
-- Your code down below

```

#### Expected results

|ProductName                     |Unit|
|--------------------------------|----|
|Mishi Kobe Niku                 |18 - 500 g pkgs.|
|Carnarvon Tigers                |16 kg pkg.|
|Sir Rodney's Marmalade          |30 gift boxes|
|Thüringer Rostbratwurst         |50 bags x 30 sausgs.|
|Côte de Blaye                   |12 - 75 cl bottles|
|Manjimup Dried Apples           |50 - 300 g pkgs.|
|Raclette Courdavault            |5 kg pkg.|


## 2. What are the top 5 most profitable products?

```sql
-- Your code down below
```

In [5]:
def simple_query (query):
    return pd.read_sql_query (query, engine)

In [20]:
query ="WITH ProductTotalRevenue AS ( SELECT products.ProductID, ProductName, SUM(Quantity) AS TotalRevenue FROM Products JOIN orderdetails ON orderdetails.ProductID = products.ProductID GROUP BY products.ProductID, ProductName) SELECT ProductID, ProductName, TotalRevenue FROM ProductTotalRevenue ORDER BY TotalRevenue DESC;"


In [7]:
simple_query(query)

ProductID                ProductName  TotalRevenue
0          31          Gorgonzola Telino         458.0
1          60          Camembert Pierrot         430.0
2          35             Steeleye Stout         369.0
3          59       Raclette Courdavault         346.0
4           2                      Chang         341.0
..        ...                        ...           ...
72         15               Genen Shouyu          25.0
73          9            Mishi Kobe Niku          20.0
74         22        Gustaf's Knäckebröd          18.0
75         45                Røgede sild          15.0
76         67  Laughing Lumberjack Lager           5.0

[77 rows x 3 columns]

#### Expected results
Alice Mutton is 6th

|ProductID                       |ProductName|TotalRevenue|
|--------------------------------|-----------|------------|
|-                             |-|-       |
|-                           |-|-       |
|-                            |-|-       |
|-                             |-|-       |
|-                              |-|-       |
|17                              |Alice Mutton|12909       |

## 3. Write a CTE that lists the top 5 categories by the number of products they have.

In [22]:
query="WITH CategoryProductCount AS (SELECT CategoryName, Count(ProductID) as ProductCount FROM categories JOIN products ON categories.CategoryID=products.CategoryID group by CategoryName) SELECT CategoryName, ProductCount FROM CategoryProductCount ORDER BY ProductCount DESC;" 
simple_query(query)

CategoryName  ProductCount
0     Confections            13
1       Beverages            12
2      Condiments            12
3         Seafood            12
4  Dairy Products            10
5  Grains/Cereals             7
6    Meat/Poultry             6
7         Produce             5

#### Expected result
|CategoryName                    |ProductCount|
|--------------------------------|------------|
|Confections                     |13          |
|Beverages                       |12          |
|Condiments                      |12          |
|Seafood                         |12          |
|Dairy Products                  |10          |


## 4. Write a CTE that shows the average order quantity for each product category.

```sql
-- Your code down below

```

In [33]:
query ="WITH table1 AS (SELECT CategoryName, avg(Quantity) AS AvgOrderQuantity FROM categories JOIN products ON products.CategoryID=categories.CategoryID  JOIN orderdetails ON orderdetails.ProductID = products.ProductID  group by CategoryName) SELECT CategoryName, AvgOrderQuantity FROM table1 order by AvgOrderQuantity DESC;"

simple_query(query)

CategoryName  AvgOrderQuantity
0      Condiments           28.2245
1  Dairy Products           26.0100
2    Meat/Poultry           25.7600
3     Confections           25.1190
4       Beverages           24.6129
5  Grains/Cereals           21.7143
6         Produce           21.6667
7         Seafood           21.5672

|CategoryName                    |AvgOrderQuantity|
|--------------------------------|----------------|
|Beverages                       |24.6129         |
|Condiments                      |28.2245         |
|Confections                     |25.1190         |
|Dairy Products                  |26.0100         |
|Grains/Cereals                  |21.7143         |
|Meat/Poultry                    |25.7600         |
|Produce                         |21.6667         |
|Seafood                         |21.5672         |


# 5. Create a CTE to calculate the average order amount for each customer.

In [35]:
query = "WITH cust_ord_table AS (SELECT customers.customerID,customerName,Quantity FROM customers JOIN orders ON customers.customerID=orders.customerID JOIN orderdetails ON orderdetails.orderID= orders.orderID GROUP BY customerID) SELECT customerID,CustomerName, Avg(Quantity) as AvgOrderAmount FROM cust_ord_table ORDER BY vgOrderAmount DESC) SELECT customers.customerID,customerName,avg(Quantity) FROM cust_ord_table ORDER DESC ;"

simple_query(query)

ProgrammingError: (pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ') SELECT customers.customerID,customerName,avg(Quantity) FROM cust_ord_table ORD' at line 1")
[SQL: WITH cust_ord_table AS (SELECT customers.customerID,customerName,Quantity FROM customers JOIN orders ON customers.customerID=orders.customerID JOIN orderdetails ON orderdetails.orderID= orders.orderID GROUP BY customerID) SELECT customerID,CustomerName, Avg(Quantity) as AvgOrderAmount FROM cust_ord_table ORDER BY vgOrderAmount DESC) SELECT customers.customerID,customerName,avg(Quantity) FROM cust_ord_table ORDER DESC ;]
(Background on this error at: https://sqlalche.me/e/20/f405)

|CustomerID                      |CustomerName|AvgOrderAmount|
|--------------------------------|------------|--------------|
|59                              |Piccolo und mehr|4014.2500     |
|73                              |Simons bistro|2444.3333     |
|62                              |Queen Cozinha|1991.6667     |
|51                              |Mère Paillarde|1673.8571     |
|71                              |Save-a-lot Markets|1407.2500     |
|76                              |Suprêmes délices|1345.8333     |
|81                              |Tradição Hipermercados|1315.6667     |
|7                               |Blondel père et fils|1174.4615     |
|89                              |White Clover Markets|1112.5000     |
|55                              |Old World Delicatessen|1079.5000     |
|20                              |Ernst Handel|1018.0000     |
|19                              |Eastern Connection|1004.8000     |
|68                              |Richter Supermarkt|976.6667      |
|72                              |Seven Seas Imports|934.0000      |
|63                              |QUICK-Stop  |908.5500      |
|25                              |Frankenversand|895.0667      |
|75                              |Split Rail Beer & Ale|854.7692      |
|65                              |Rattlesnake Canyon Grocery|838.2727      |
|52                              |Morgenstern Gesundkost|754.0000      |
|9                               |Bon app''   |750.4286      |
|37                              |Hungry Owl All-Night Grocers|733.5714      |
|35                              |HILARIÓN-Abastos|722.8333      |
|33                              |GROSELLA-Restaurante|690.0000      |
|34                              |Hanari Carnes|681.0000      |
|15                              |Comércio Mineiro|677.5000      |
|23                              |Folies gourmandes|672.3333      |
|10                              |Bottom-Dollar Marketse|664.0000      |
|31                              |Gourmet Lanchonetes|641.0000      |
|60                              |Princesa Isabel Vinhoss|628.5000      |
|5                               |Berglunds snabbköp|601.5556      |
|46                              |LILA-Supermercado|564.7692      |
|44                              |Lehmanns Marktstand|545.7500      |
|30                              |Godos Cocina Típica|515.0000      |
|3                               |Antonio Moreno Taquería|504.0000      |
|88                              |Wellington Importadora|503.8333      |
|47                              |LINO-Delicateses|500.0000      |
|56                              |Ottilies Käseladen|500.0000      |
|87                              |Wartian Herkku|492.5833      |
|24                              |Folk och fä HB|480.2222      |
|49                              |Magazzini Alimentari Riuniti|464.2857      |
|86                              |Die Wandernde Kuh|441.9091      |
|41                              |La maison d''Asie|437.8182      |
|8                               |Bólido Comidas preparadas|416.0000      |
|83                              |Vaffeljernet|411.0000      |
|14                              |Chop-suey Chinese|374.8333      |
|80                              |Tortuga Restaurante|357.4167      |


# 6. Sales Analysis with CTEs

Assume we have the Northwind database which contains tables like Orders, OrderDetails, and Products. Create a CTE that calculates the total sales for each product in the year 1997.

```sql
-- Your code down below

 ```

#### Expected result

|ProductName                     |TotalSales|
|--------------------------------|----------|
|Gnocchi di nonna Alice          |173       |
|Tourtière                       |126       |
|Geitost                         |119       |
|Chang                           |115       |
|Raclette Courdavault            |115       |
|Sirop d'érable                  |106       |
|Vegie-spread                    |100       |
|Côte de Blaye                   |99        |
|Alice Mutton                    |97        |
|Steeleye Stout                  |95        |
|Sir Rodney's Scones             |92        |
|Pavlova                         |86        |
|Zaanse koeken                   |85        |
|Fløtemysost                     |75        |
|Tarte au sucre                  |75        |
